### ISSUE TITLE
Error support

### ISSUE THREAD

**Issue by vikram7** [fotinakis/jsonapi-serializers#22]
Currently, I don't think that errors are supported. For example, the JSON API provides for the following structure for error responses:

```javascript
{
  "errors": [
    {
      "source": { "pointer": "/data/attributes/first-name" },
      "title": "Invalid Attribute",
      "detail": "First name must contain at least three characters."
    },
    {
      "source": { "pointer": "/data/attributes/first-name" },
      "title": "Invalid Attribute",
      "detail": "First name must contain an emoji."
    }
  ]
}
```

From: [JSON API examples](http://jsonapi.org/examples/)

The current `serialize` methods don't support cases where responses are failures. Are there plans to support how errors are handled by `jsonapi-serializers`?

**Comment by fotinakis** [comment]
@vikram7 we should probably support that, yes. I haven't put a lot of thought into how that interface should look—any ideas? Just `serialize(obj, error=true)` or something?

It's a bit complicated because jsonapi-serializers is not directly tied to ActiveRecord models, and when constructing those error responses we'd be assuming a lot about about where errors are stored on objects (ie. for ActiveModel/ActiveRecord we could read from the `.errors` or `.valid?` attributes). In this case it might be worth that coupling to keep the public interface nice and support the common use case, but also it'd be nice to have somewhat generic error support in this gem.

**Comment by vikram7** [comment]
@fotinakis I see where you are coming from. I too think there is value in allowing for some kind of generic error support. Trying to figure out a way to do it without coupling with ActiveModel/ActiveRecord would be nice. What do you think about the following approach?:

`serialize(obj, errors)`

`errors` here would be an hash of errors in some prescribed format:

`errors = { :attr1 => ["can't be blank"], :attr2 => ["can't be blank"], . . . }`

The `serialize` method, if errors are optionally passed to it would then return the following:

```javascript
{
  "errors": [
    {
      "source": { "pointer": "/data/attributes/attr1" },
      "title": "Invalid Attribute",
      "detail": "can't be blank"
    },
    {
      "source": { "pointer": "/data/attributes/attr2" },
      "title": "Invalid Attribute",
      "detail": "can't be blank"
    }
  ]
}
```

The onus, here, would be on the user to provide the serializer an error hash in the correct format `jsonapi-serializers` prescribes.

**Comment by mwpastore** [comment]
:+1:

**Comment by tobsch** [comment]
+1

**Comment by matheusca** [comment]
Hi!

I think couple any framework or lib to creating errors is shouldn't useful for me, for example I am not use Rails, it not just for me, because many people do not use too. What I would do in this case it'd be to define a interface that developer could be do. For instance:

```ruby
class Error
  include JSONAPI::Error

  def initialize(object)
    @object = object
  end

  def errors
    @object.errors.each do |error|
       {code: error.code, source: error.source}
     end
  end
end

# Using
Error.new(entity).serialize
# Retuning jsonapi error as described in spec.
```

In this case could be define how the hash error would be like.

What do you think guys?

Thanks

**Comment by arvanasse** [comment]
I tend to agree with @matheusca.  Since every attribute serialized for the errors collection is optional it is not essentially different than serializing any other collection.  The only difference between serializing a (root) object or collection and serializing an error collection is the key used at the root level of the json (data vs errors).

With that in mind, perhaps the JSONAPI::Serializer module could include a mechanism for specifying the root key with the default value being 'data'.  An implementation desiring an error serializer could request 'errors' for their error serializer and then provide a list of attributes corresponding to their preferred error attributes.  It is likely that an error presenter would be necessary to translate ActiveModel::Errors collections (and maybe a basic implementation could be included via railtie as an example?).

**Comment by hspazio** [comment]
Hi @fotinakis 
I would like to resume this thread if anybody is still interested. I've been looking at serializing errors recently and I was thinking of this kind of solution.

```ruby
class ErrorSerializer
  # create a different type of serializer that does not 
  # conflict with the original Serializer class
  include JSONAPI::ErrorSerializer
end

# The ErrorSerializer just need to work with an object 
# that responds to :status, :source and :detail messages. 
# Perhaps in our application we can have a class like 
# below to generate an error object.
class ApplicationError
  attr_accessor :status, :source, :detail

  def initialize(params = {})
    @status = params[:status]
    @source = params[:source]
    @detail = params[:detail]  
  end

  # explicit serializer discovery ensures all instances of 
  # this class are handled using the correct serializer
  def jsonapi_serializer_class_name
    'ErrorSerializer'
  end  
end

# in order to serialize an error we need to instantiate the error object
error = ApplicationError.new(params)

# ... and serialize using the existing class method if possible
JSONAPI::Serializer.serialize(error)
# or 
JSONAPI::ErrorSerializer.serialize(error)

# in case of collection of errors we could do the same
JSONAPI::Serializer.serialize(errors, is_collection: true)
# or
JSONAPI::ErrorSerializer.serialize(errors, is_collection: true)
```
I would very much appreciate any feedback on this and perhaps I could submit a PR soon.

**Comment by elkelk** [comment]
I think I'd prefer @vikram7's approach. It seems like a good first step to at least allow the user to specify an "errors" key in the options (similar to the way one can specify "meta").

**Comment by fotinakis** [comment]
Fixed by #63 and released in v0.7.0.
